### ISSUE TITLE
CDM Serialization Specification

### ISSUE THREAD

**Issue by chrisisla** [finos/common-domain-model#3236]
## Background

To meet its promise of interoperability, CDM requires a defined standard for the exchange of information. Unfortunately, after a detailed analysis, the community determined that the current JSON format had gaps that prevented it from becoming that standard. In November 2023, the SWG approved recommendations made by a TAWG task force for a phased approach to changes to the data standard.

As part of its commitment to promoting the CDM as a standard, ISLA is funding the first phase of this work, working in conjunction with CLOUDRISK, FT Advisory, ISDA and REGnosys. The following provides an overview of the plan for the changes.

The high level goals for this phase of the Serialization project were to focus on:

- **interoperability**: users of the same version of the model should be able to exchange data regardless of their programming language
- **completeness**: ability to represent the entire model
- **readability**: serialized data should be human readible
- **compactness**: serialized data should be as compact as possible

The design, once accepted by the community, will move into development with a targeted release of CDM version 7.

## Principles

When evaluating serialization design alternatives the followed ordered list of principles served as reference:

- **object generation**: serialized data should facilitate creation of Rune defined objects including enabling the "language's" polymorphic inheritance
- **model conformity**: to the fullest extent possible, the serialized data should directly conform to the model.  The reader has no obligation to keep fields that it does not recognise
- **error reporting**: report all failures
- **atomic types serialisation**: data types as dates, times and others should be serialized as far as possible using well established standards/formats as for example ISO.
- **ease of integration**: to the extent possible, it should be easy to keep serialization in sync with model versions

Some may not be absolute and the final design may have had to make compromises on the extent to which it meets the principles.

## Proposal

The CDM object serialization will be enhanced to now include the following new _special_ attributes.

| New | Example | Description | 
| --- | --- | --- | 
|  `@model` | `"@model": "cdm"` | Model short name, located at the top of the JSON |
|  `@version` | `"@version": "1.2.3"` | Model version, located at the top of the JSON |
|  `@type` | `"@type": "cdm.event.common.TradeState"` | Fully Qualified Type Name, located at the top of the JSON |
|  `@data`| `"@data": "54930084UKLVMY22DS16"` | Replaces `value` annotation | 
|  `@key` | `"@key": "abcd1234"` | Replaces `globalKey` metadata annotation |
|  `@ref` | `"@ref": "abcd1234"` | Replaces `globalReference` metadata annotation |
|  `@key:external` | `"@key:external": "my-external-key"` | Replaces `externalKey` metadata annotation  | 
|  `@ref:external` | `"@ref:external": "my-external-key"` | Replaces `externalReference` metadata annotation  |
|  `@key:scoped` | `"@key:scoped": "my-scoped-key"` | Replaces `scope` and `location` annotations  | 
|  `@ref:scoped` | `"@ref:scoped": "my-scoped-key"` | Replaces `scope` and `address` annotations |
|  `@scheme` | `"@scheme": "http://www.fpml.org/coding-scheme/external/iso17442"` | Replaces `scheme` metadata annotation |

When serializing:
- _special_ attributes will always be at the top of a node
- null _special_ attributes will no longer be serialized
- internal keys that are not referenced will not be serialized; external keys will always be serialized

When deserializing:
- JSON will be validated for structure and types, not content
- attributes that do not conform to the current version of the model will be discarded

In summary the new serialized format:
- has stronger **alignment** to model definitions
- improves **readability** 
- is more **compact**
- provides a more **consistent** machine-readable representation
- facilitates **easier and faster** machine processing. 

For more details of the proposed structure and processing please refer to the full design document available here: 
[Serialization Specification v0.9.pdf](https://github.com/user-attachments/files/17680613/Serialization.Specification.v0.9.pdf)

## Feedback

If you have any questions on the design then please add them as new comments to this issue.

**Comment by chrisisla** [comment]
The design will be presented at the TAWG meeting on November 14th (details of the meeting can be found here: #3212). 

The community will then have 3 weeks to review the design and pose any questions or comments that they may have. The expectation is that objections, if any, should be concrete, material and include alternative approaches so that a final review can address the pros and cons.

At the following meeting of the TAWG on December 5th any outstanding questions on the design will be answered, with a view to the final design being accepted at that meeting.

**Comment by ahutusoru123** [comment]
Question regarding the @data notation - clarified during the call that @data should be used when a primitive value (end leaf) has some meta data attached to it. Conclusion was: adding a larger CDM object example would add clarity on proposed usage.

**Comment by chrisisla** [comment]
As requested, a please find a larger example of the new serialization structure. 
[Serialization Comparison.pdf](https://github.com/user-attachments/files/17906160/Serialization.Comparison.pdf)

cc: @ahutusoru123

**Comment by davis-io** [comment]
Question - looking at the section that shows a side-by-side comparison, should "@data" be used rather than "value" on the right-hand side on line 41? If not, what's the difference between this and the use of "@data" on line 64?

**Comment by chrisisla** [comment]
Hi @davis-io , you are correct, line 41 should be `@data` as well, I have corrected the document.
[Serialization Comparison.pdf](https://github.com/user-attachments/files/17916921/Serialization.Comparison.pdf)

**Comment by brianlynn2** [comment]
Overall evaluation:  This is a very good first step in defining a design direction for serialization.  Looks excellent in terms of defining the serialization format - compact and comprehensive.  It doesn’t fully define how serialization and deserialization will be used in practice; this will need to be addressed in follow-on work. 

Comment on documentation of the spec: 
•	It would be helpful to have an additional column or columns on the table of special attributes to specify whether the attributes are applicable for the root of the document or within a document (or both).

Missing/Open Questions
•	There are some topics not covered by the serialization document that will need to be covered in subsequent design work.  It’s difficult to fully assess the serialization design without these types of questions being answered.
•	How will this be accessed by a CDM user?  Will there be a Rune function that allows a CDM app to load (deserialize) or save (serialize and write out) a CDM object?   Will there be a defined mechanism for doing this in Java, Python, other languages?  How will those work?
•	What will happen during deserialization if an enumerated value (i.e. a value represented by a CDM enumeration) has an invalid code in the serialized data?
•	Will there be a mechanism provided for triggering data validation for a newly loaded (deserialized) object, e.g. string length validation?  If this is out of scope of the deserialization spec, is there a way for a user to trigger it otherwise?   An example of how to do that would be helpful.
•	What if any checking is envisaged for the @version attribute?  For instance, what will happen during deserialization  if the @version attribute in the file being loaded is lower (older) than the version of CDM that is being run?  (Does that depend if the major version is the same?)  What about if it is newer than the version of CDM that is being run? (Does it depend on whether the major version matches?)  Will there be any mechanism for constraining the supported versions for a given operation?  (e.g. on a load specifying that the minimum version is xx.xx).   I know that this spec isn’t addressing version interoperability/migration, but it should be possible to write data using CDM version 6.0 and read it using 6.1.   And probably vice versa as long as none of the new data in 6.1 are used.

**Comment by plamen-neykov** [comment]
One quick comment about the enum validation - they should be treated as basic (atomic) types, and should be validated as part of the basic validation (step `Verify Types`) - see the flow-chart in section 5.3.1 of the [Serialization Specification v0.9.pdf](https://github.com/user-attachments/files/17680613/Serialization.Specification.v0.9.pdf).

**Comment by plamen-neykov** [comment]
Another quick comment on the serialisation/deserialisation of Rune objects (this is not a CDM specific functionality, but is a Rune level facility) - it has been discussed to include in the respective runtime implementations (java, python) identically defined function calls (function and parameter names) for serializing and deserializing Rune objects.

**Comment by brianlynn2** [comment]
@plamen-neykov does 5.3.1 mean that an illegal value of an enumeration would " provide a warning and discard any attributes that do not conform to the model."   I guess so.

It would indeed be helpful if there were standardized function calls for serializing and deserializing Rune objects, ideally also accessible from within Rune itself.  We would want to use this functionality for loading reference data in Rune object format, for instance.

**Comment by plamen-neykov** [comment]
> @plamen-neykov does 5.3.1 mean that an illegal value of an enumeration would " provide a warning and discard any attributes that do not conform to the model." I guess so.
> 
@brianlynn2 errors in the **Basic Validation** step would cause the ingestion to stop and issue an error - no model will be constructed if an error occurs in this step - @chrisisla @minesh-s-patel @dschwartznyc please comment if your understanding differs.

Warnings will be emitted in the **Model Clean Up** and **Model Validation** steps.

On 
> It would indeed be helpful if there were standardized function calls for serializing and deserializing Rune objects, ideally also accessible from within Rune itself. We would want to use this functionality for loading reference data in Rune object format, for instance.

my personal opinion is that it should be possible (easy) to expose the serialisation/deserialisation functions to Rune and make them invokable from Rune code. Alas, I believe this would be something well outside of the scope of the serialisation project.

**Comment by chrisisla** [comment]
At the meeting of the TAWG on December 5th the Serialization Specification held in this Issue was formally agreed. 

Please refer to #3219 for details.

**Comment by manel-martos** [comment]
We feel mostly comfortable with the proposal but we'd like to share some concerns observed during our review, particularly on extended models.

The proposal introduces different approaches for representing basic types based on the presence or absence of annotations. This inconsistency complicates data extraction from serialized instances, particularly when dealing with changes in the metadata or model definitions.

For example, consider the following:
`
type QuantitySchedule:
  value number (0..1)
  unit UnitType (0..1)
`

This will be serialized as:
`
{
  "value": 150000,
  "unit": {
    "financialUnit": "Share"
  }
}
`

If the model is updated with metadata annotations as follows:
`
type QuantitySchedule:
  value number (0..1)
  [metadata id]
  unit UnitType (0..1)
`

The serialized representation would change to:
`
{
  "value": {
    "@key": "abc",
    "@data": 150000
  },
  "unit": {
    "financialUnit": "Share"
  }
}
`

**Observations:**
Even though we agree it is an improvement, this change still introduces additional layers of nesting and complexity, which impacts downstream processes that extract or process data from serialized objects.

- Adding metadata annotations results in the value being serialized as a nested object (@key and @data) rather than a direct representation. This necessitates additional parsing logic to handle both the annotated and non-annotated serialization formats.
- The mixed serialization style makes it difficult for consumers of the model to create reusable deserialization logic.
- Tools or scripts processing the serialized data must account for multiple potential structures for the same type.

Sections 5 and 7 doesn't seem to clarify this so we'd ask whether you already considered this or other future improvements or you do not consider it as a concern at all. 

Thanks!

**Comment by tomhealey-icma** [comment]
I have some comments on Brian's questions:

• How will this be accessed by a CDM user? Will there be a Rune function that allows a CDM app to load (deserialize) or save (serialize and write out) a CDM object? Will there be a defined mechanism for doing this in Java, Python, other languages? How will those work?

_Running serialization and deserialization is currently performed through the Regnosys RosettaObjectMapper which is a separate library and is language dependent. So, Bryan’s question raises other questions, will the new serialization be supported by Regnosys and if not, who? Can it move to FINOS?_

• What will happen during deserialization if an enumerated value (i.e. a value represented by a CDM enumeration) has an invalid code in the serialized data?

_My experience is that it’s not possible to use the RosettaObjectMapper and generate serialized output if the enumerations do not compile. If it’s a question of a new enum value or a change, that would mean a version change in the CDM and while one application may compile with a new enum, it would not be possible to create an object in a separate application if the enum does not exist._

• Will there be a mechanism provided for triggering data validation for a newly loaded (deserialized) object, e.g. string length validation? If this is out of scope of the deserialization spec, is there a way for a user to trigger it otherwise? An example of how to do that would be helpful.

_There is no data validation that would not have been performed by the model during compile time or at runtime. Again, another question would be, will the new serialization support custom data types added to the runtime model. Given the very limited number of data types and validation I’ve added new data types and runtime functions that work in the model and serialization process so it’s important to know if this will continue to be supported._

• What if any checking is envisaged for the [@Version](https://github.com/Version) attribute? For instance, what will happen during deserialization if the [@Version](https://github.com/Version) attribute in the file being loaded is lower (older) than the version of CDM that is being run? (Does that depend if the major version is the same?) What about if it is newer than the version of CDM that is being run? (Does it depend on whether the major version matches?) Will there be any mechanism for constraining the supported versions for a given operation? (e.g. on a load specifying that the minimum version is xx.xx). I know that this spec isn’t addressing version interoperability/migration, but it should be possible to write data using CDM version 6.0 and read it using 6.1. And probably vice versa as long as none of the new data in 6.1 are used.

_Versioning at the top level only is problematic. It’s possible and necessary to serialize at the type level (parties, party roles, business centers, etc). A scenario where a new top level version is not backward compatible would restrict serialization at the type level where the types are still compatible. Versioning at type level objects would make versioning more useful otherwise a version check should be done separately by the application._ 

_Lastly, the TAWG serialization proposal defined a set of required, optional and nice-to-have features. I don’t see those anywhere in this proposal._
